cmd/optimizer-nri-plugin/main.go

```diff
@@ -0,0 +1,327 @@
+	return []cli.Flag{
+		&cli.StringFlag{
+			Name:        "name",
+			Usage:       "plugin name to register to NRI",
+			Destination: &args.PluginName,
+		},
+		&cli.StringFlag{
+			Name:        "idx",
+			Usage:       "plugin index to register to NRI",
+			Destination: &args.PluginIdx,
+		},
+		&cli.StringFlag{
+			Name:        "events",
+			Value:       defaultEvents,
+			Usage:       "the events that containerd subscribes to",
+			Destination: &args.PluginEvents,
+		},
+		&cli.StringFlag{
+			Name:        "log-file",
+			Usage:       "the log file path for optimizer client",
+			Destination: &args.Config.LogFile,
+		},
+		&cli.StringFlag{
+			Name:        "server-path",
+			Value:       defaultServerPath,
+			Usage:       "the path of optimizer server binary",
+			Destination: &args.Config.ServerPath,
+		},
+		&cli.StringFlag{
+			Name:        "persist-dir",
+			Value:       defaultPersistDir,
+			Usage:       "the directory to persist accessed files list for container",
+			Destination: &args.Config.PersistDir,
+		},
+		&cli.IntFlag{
+			Name:        "timeout",
+			Value:       0,
+			Usage:       "the timeout to kill optimizer server, 0 to disable it",
+			Destination: &args.Config.Timeout,
+		},
+		&cli.BoolFlag{
+			Name:        "overwrite",
+			Usage:       "whether to overwrite the existed persistent files",
+			Destination: &args.Config.Overwrite,
+		},
+	}
+}
+
+func NewPluginFlags() *Flags {
+	var args PluginArgs
+	return &Flags{
+		Args: &args,
+		F:    buildFlags(&args),
+	}
+}
+
+type plugin struct {
+	stub stub.Stub
+	mask stub.EventMask
+}
+
+var (
+	cfg                  PluginConfig
+	log                  *logrus.Logger
+	logFile              *os.File
+	_                    = stub.ConfigureInterface(&plugin{})
+	globalFanotifyServer = make(map[string]*fanotify.Server)
+)
+
+const (
+	imageNameLabel = "io.kubernetes.cri.image-name"
+)
+
+func (p *plugin) Configure(config, runtime, version string) (stub.EventMask, error) {
+	log.Infof("got configuration data: %q from runtime %s %s", config, runtime, version)
+	if config == "" {
+		return p.mask, nil
+	}
+
+	oldCfg := cfg
+	tree, err := toml.Load(config)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse TOML: %w", err)
+	}
+	if err := tree.Unmarshal(&cfg); err != nil {
+		return 0, err
+	}
+
+	p.mask, err = api.ParseEventMask(cfg.Events...)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse events in configuration: %w", err)
+	}
+
+	if cfg.LogFile == "" {
+		cfg.LogFile = defaultLogFile
+	}
+	if cfg.LogFile != oldCfg.LogFile {
+		file, err := redirectLoggerToFile(cfg.LogFile)
+		if err != nil {
+			return 0, err
+		}
+		if logFile != nil {
+			logFile.Close()
+		}
+		logFile = file
+	}
+	log.Infof("configuration: %#v", cfg)
+
+	return p.mask, nil
+}
+
+func (p *plugin) StartContainer(pod *api.PodSandbox, container *api.Container) error {
+	dump("StartContainer", "pod", pod, "container", container)
+
+	imageName := GetImageName(container.Annotations)
```

Comment:
We should validate the image. Containerd already provides a package to validate image name.

Reply:
Done.